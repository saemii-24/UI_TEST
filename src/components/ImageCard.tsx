import React, { useRef, useEffect, useState } from "react";
import Image from "next/image";
import type { ImageListProps } from "@/types/type";
import { mockImageData } from "@/test/mockdata";

const ImageCard = ({
  imageList = mockImageData[0],
}: {
  imageList: Partial<ImageListProps>;
}) => {
  const imageCardRef = useRef<HTMLDivElement>(null);
  const [cardWidth, setCardWidth] = useState(0);

  // 이미지 비율에 맞는 height 계산
  const calculateHeight = (width: number) => {
    return (width * (imageList.imageHeight ?? 1)) / (imageList.imageWidth ?? 1);
  };

  useEffect(() => {
    if (imageCardRef.current) {
      setCardWidth(imageCardRef.current.offsetWidth);
    }
    // 리사이즈 이벤트에 대응
    const handleResize = () => {
      if (imageCardRef.current) {
        setCardWidth(imageCardRef.current.offsetWidth);
      }
    };
    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  }, []);

  if (!imageList.webformatURL) return null;

  return (
    <div data-testid='image-card' ref={imageCardRef}>
      <div className='relative w-full max-w-[400px] cursor-pointer overflow-hidden rounded-xl before:absolute before:z-10 before:size-full before:opacity-40 before:transition-all before:duration-150 hover:before:bg-black'>
        <Image
          className=' bg-gray-200'
          src={imageList.webformatURL}
          alt={`${imageList.id} image-thumbnail ${imageList.user}`}
          width={cardWidth}
          height={calculateHeight(cardWidth)}
          sizes={`${cardWidth}px`}
          priority
        />
      </div>
      {imageList.user &&
        imageList.user.length > 0 &&
        imageList.userImageURL &&
        imageList.userImageURL.length > 0 && (
          <div className='mt-2 flex items-center gap-2'>
            <div className='relative size-10 overflow-hidden rounded-full'>
              <Image
                src={imageList.userImageURL}
                alt={`${imageList.id} image-user ${imageList.user}`}
                fill
                style={{ objectFit: "cover" }}
                priority
              />
            </div>
            <div className='text-base font-semibold'>Photo by. {imageList.user}</div>
          </div>
        )}
    </div>
  );
};

export default ImageCard;
