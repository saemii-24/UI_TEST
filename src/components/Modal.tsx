"use client";
import React, { useEffect, useState, useRef } from "react";
import useModalImageId from "@/store/modalImageIdStore";
import Image from "next/image";
import { VscChromeClose } from "react-icons/vsc";
import { RiHeart3Line, RiHeart3Fill } from "react-icons/ri";
import cn from "classnames";
import Button from "./Button";

const Modal = () => {
  const { modalImage, setModalImage } = useModalImageId();
  const [isLike, setIsLike] = useState<boolean>(false);

  const modalRef = useRef<HTMLDivElement | null>(null); // 모달 흰색 부분
  const backgroundRef = useRef<HTMLDivElement | null>(null); // 모달 배경

  // 모달 외부 클릭 시 모달 닫기
  useEffect(() => {
    const clickOutSide = (event: MouseEvent) => {
      const eventTarget = event.target;
      // 클릭된 곳이 모달 외부인 경우 모달을 닫음
      if (eventTarget === backgroundRef.current) {
        setModalImage(undefined);
      }
    };

    window.addEventListener("mousedown", clickOutSide);

    return () => {
      window.removeEventListener("mousedown", clickOutSide);
    };
  }, [setModalImage]);

  // 모달 열릴 때 스크롤 막고 레이아웃 이동 방지
  useEffect(() => {
    const preventScroll = () => {
      if (typeof window !== "undefined") {
        const scrollBarWidth = window.innerWidth - document.documentElement.clientWidth;
        document.body.style.overflow = "hidden"; // 스크롤 방지
        document.body.style.paddingRight = `${scrollBarWidth}px`; // 스크롤바 너비만큼 padding-right 추가함
      }
    };

    const enableScroll = () => {
      if (typeof window !== "undefined") {
        document.body.style.overflow = "";
        document.body.style.paddingRight = "";
      }
    };

    // 모달이 열리면 preventScroll, 닫히면 enableScroll
    if (modalImage) {
      preventScroll();
    } else {
      enableScroll();
    }

    return () => {
      enableScroll(); // 컴포넌트 언마운트 시 스크롤 복원
    };
  }, [modalImage]);

  // 모달이 열리지 않아야 하는 상황
  if (!modalImage || !modalImage.webformatURL) {
    setModalImage(undefined);
    return null;
  }

  return (
    <div
      ref={backgroundRef}
      className='fixed inset-0 z-[9999] h-screen w-full bg-black/30'
    >
      <div
        ref={modalRef}
        className='relative left-1/2 top-1/2 h-[70vh] w-[90%] max-w-[500px] -translate-x-1/2 -translate-y-1/2 overflow-hidden rounded-2xl bg-white shadow-lg'
      >
        <div className='absolute right-4 top-4 z-[10000] cursor-pointer rounded-full bg-white  p-2 text-2xl font-bold text-black'>
          <VscChromeClose
            className='text-black'
            onClick={() => setModalImage(undefined)}
          />
        </div>
        <div className={cn("relative w-full h-[55%] flex-col")}>
          <Image
            src={modalImage.webformatURL as string}
            sizes='(max-width: 768px) 100vw, (max-width: 1200px) 500px'
            alt={`${modalImage.id} image-thumbnail ${modalImage.user}`}
            priority
            fill
            objectFit='cover'
            className=' bg-gray-200'
          />
        </div>
        <div className='flex size-full h-[45%] flex-col  px-5 pb-5 pt-4'>
          <div className='flex items-center justify-between gap-3'>
            {modalImage.user &&
              modalImage.user.length > 0 &&
              modalImage.userImageURL &&
              modalImage.userImageURL.length > 0 && (
                <div className='mt-2 flex flex-1 items-center gap-2'>
                  <div className='relative size-16 overflow-hidden rounded-full'>
                    <Image
                      src={modalImage.userImageURL}
                      alt={`${modalImage.id} image-user ${modalImage.user}`}
                      width='80'
                      height='80'
                      style={{ objectFit: "cover" }}
                      priority
                    />
                  </div>
                  <div className='flex-col'>
                    {/* 사진 촬영자 이름 */}
                    <div className='flex items-center justify-between'>
                      <div className='text-xl font-semibold'>
                        Photo by. {modalImage.user}
                      </div>
                      <div
                        onClick={() => {
                          setIsLike(!isLike);
                          if (modalImage.likes) {
                            modalImage.likes = isLike
                              ? modalImage.likes - 1
                              : modalImage.likes + 1;
                            setModalImage(modalImage);
                          }
                        }}
                        className='mt-2 flex cursor-pointer items-center gap-1'
                      >
                        {isLike ? (
                          <RiHeart3Fill className='text-lg text-red-500' />
                        ) : (
                          <RiHeart3Line className='text-lg text-red-500' />
                        )}
                        {/* <div>{modalImage.likes}</div> */}
                      </div>
                    </div>
                    {/* 이미지 태그 */}
                    <div className=' mt-1 flex gap-2 text-sm text-gray-600'>
                      {modalImage.tags?.split(", ").map((item, index) => {
                        return <div key={index}>#{item.replace(/\s+/g, "_")}</div>;
                      })}
                    </div>
                    {/* 이미지 좋아요 버튼 */}
                    {/* <div
                      onClick={() => {
                        setIsLike(!isLike);
                        if (modalImage.likes) {
                          modalImage.likes = isLike
                            ? modalImage.likes - 1
                            : modalImage.likes + 1;
                          setModalImage(modalImage);
                        }
                      }}
                      className='mt-2 flex cursor-pointer items-center gap-1'
                    >
                      {isLike ? (
                        <RiHeart3Fill className='text-lg text-red-500' />
                      ) : (
                        <RiHeart3Line className='text-lg text-red-500' />
                      )}
                      <div>{modalImage.likes}</div>
                    </div> */}
                  </div>
                </div>
              )}
          </div>

          {/* <Button
            rounded='full'
            width='full'
            className=''
            onClick={() => {
              setIsLike(!isLike);
              if (modalImage.likes) {
                modalImage.likes = isLike ? modalImage.likes - 1 : modalImage.likes + 1;
                setModalImage(modalImage);
              }
            }}
          >
            <div className='mt-2 flex cursor-pointer items-center gap-1'>
              {isLike ? (
                <RiHeart3Fill className='text-lg text-red-500' />
              ) : (
                <RiHeart3Line className='text-lg text-red-500' />
              )}
              <div>{modalImage.likes}</div>
            </div>
          </Button> */}
          <Button rounded='full' width='full' color='blue' className='mt-auto'>
            다운로드
          </Button>
        </div>
      </div>
    </div>
  );
};

export default Modal;
