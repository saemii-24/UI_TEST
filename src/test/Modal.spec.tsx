import ImageCard from "@/components/ImageCard";
import { cleanup, fireEvent, getByTestId, render, waitFor } from "@testing-library/react";
import { mockImageData } from "./mockdata";
import userEvent from "@testing-library/user-event";
import Modal from "@/components/Modal";

describe("Modal 테스트", () => {
  it("사용자가 클릭 후 모달이 DOM에 추가되는지 확인한다.", async () => {
    // ImageCard 컴포넌트를 렌더링하고, 모달은 아직 렌더링되지 않음
    const { queryByTestId, getByTestId } = render(
      <ImageCard imageList={mockImageData[0]} />,
    );

    // 모달이 DOM에 없다면, 모달은 ImageCard 내부에서 조건부로 렌더링되므로 처음엔 존재하지 않음
    const modal = queryByTestId("image-modal"); //query를 사용할 경우 찾지 못하더라도 에러를 반환하지 않음
    expect(modal).not.toBeInTheDocument();

    // 이미지 카드 클릭 이벤트 발생
    const imageCard = getByTestId("image-card");
    fireEvent.click(imageCard); // 클릭 이벤트 트리거

    // 클릭 후 모달이 DOM에 추가되었는지 확인
    await waitFor(() => {
      expect(screen.queryByTestId<HTMLDivElement>("image-modal")).toBeInTheDocument();
    });
  });
  it("사용자가 모달 Close 버튼을 클릭할 때 모달이 닫힌다.", () => {});
  it("사용자가 ESC키를 눌렀을 때 모달이 닫힌다.", () => {});
  it("모달 외부를 클릭하면 모달이 닫힌다.", () => {});
  it("사용자가 좋아요를 눌렀을 때 색상이 적절히 변경된다.", () => {});
  it("사용자가 다운로드 버튼을 클릭하면 해당 url의 새로운 창이 열린다.", () => {});
  it("사용자가 공유 버튼을 클릭하면 해당 url이 복사된다.", () => {});
});
