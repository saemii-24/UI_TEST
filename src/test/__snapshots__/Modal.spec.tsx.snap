// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal 테스트 > 사용자가 클릭 후 모달이 DOM에 추가되는지 확인한다. 1`] = `
<DocumentFragment>
  <div
    data-testid="image-card"
  >
    <div
      class="relative w-full max-w-screen-md cursor-pointer overflow-hidden rounded-xl before:absolute before:z-10 before:size-full before:opacity-40 before:transition-all before:duration-150 hover:before:bg-black"
    >
      <img
        alt="123456 image-thumbnail 신짱구"
        class="bg-gray-200"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="0"
        src="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2FXlemnq7yDxo7UySM67TdWk7bLD6YlH1DVCcEsDsDUBzA1YObHzAkw2HNhT68TsBWNpN7W80B8JhLxY_MFWd0Yw.webp&w=16&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2FXlemnq7yDxo7UySM67TdWk7bLD6YlH1DVCcEsDsDUBzA1YObHzAkw2HNhT68TsBWNpN7W80B8JhLxY_MFWd0Yw.webp&w=16&q=75 1x"
        style="color: transparent; width: 0px; height: 0px;"
        width="0"
      />
    </div>
    <div
      class="mt-2 flex items-center gap-2"
    >
      <div
        class="relative size-10 overflow-hidden rounded-full"
      >
        <img
          alt="123456 image-user 신짱구"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="40"
          src="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=96&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=96&q=75 2x"
          style="color: transparent; object-fit: cover;"
          width="40"
        />
      </div>
      <div
        class="text-base font-semibold"
      >
        Photo by. 신짱구
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal 테스트 > 사용자가 클릭 후 모달이 DOM에 추가되는지 확인한다. 2`] = `
<DocumentFragment>
  <div
    data-testid="image-card"
  >
    <div
      class="relative w-full max-w-screen-md cursor-pointer overflow-hidden rounded-xl before:absolute before:z-10 before:size-full before:opacity-40 before:transition-all before:duration-150 hover:before:bg-black"
    >
      <img
        alt="123456 image-thumbnail 신짱구"
        class="bg-gray-200"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="0"
        src="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2FXlemnq7yDxo7UySM67TdWk7bLD6YlH1DVCcEsDsDUBzA1YObHzAkw2HNhT68TsBWNpN7W80B8JhLxY_MFWd0Yw.webp&w=16&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2FXlemnq7yDxo7UySM67TdWk7bLD6YlH1DVCcEsDsDUBzA1YObHzAkw2HNhT68TsBWNpN7W80B8JhLxY_MFWd0Yw.webp&w=16&q=75 1x"
        style="color: transparent; width: 0px; height: 0px;"
        width="0"
      />
    </div>
    <div
      class="mt-2 flex items-center gap-2"
    >
      <div
        class="relative size-10 overflow-hidden rounded-full"
      >
        <img
          alt="123456 image-user 신짱구"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="40"
          src="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=96&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fi.namu.wiki%2Fi%2Fski_BSi3UIzyfgirmri8gUnoBpU08nLxs-af7DxS5qKf-QeP_rDP_6eh05bqtWueF6ifI3B3jadmYsbeEVm5ThTrokCDiByt8pIrv-TC--HuVEwU72DjHEOarB5LLPDKacLuM7xEvkMMH1evfsjEUw.webp&w=96&q=75 2x"
          style="color: transparent; object-fit: cover;"
          width="40"
        />
      </div>
      <div
        class="text-base font-semibold"
      >
        Photo by. 신짱구
      </div>
    </div>
  </div>
</DocumentFragment>
`;
